[KMSv2] use encDEK, keyID and annotations to generate cache key

aramase opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->
Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
It is possible for a KMSv2 plugin to return a static value as Ciphertext
and store the actual encrypted DEK in the annotations. In this case,
using the encDEK will not work. Instead, we are now using a combination
of the encDEK, keyID and annotations to generate the cache key.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aramase commented:
/sig auth
/triage accepted
/priority important-soon

aramase commented:
/assign @enj

enj commented:
/milestone v1.27

valaparthvi commented:
Hi @aramase, I am from the Bug Triage Team for K8s release v1.27. I see this PR is in actively worked up, so I assume it is on track for this release.

However, please note that Code Freeze begins on March 15, after which only critical and release blocker PR/issues will be accepted. Please make sure this PR goes in by March 15.

Do let us know if you need help in getting this PR reviewed, we can work something up :)

dims commented:
@aramase please rebase!

aramase committed:
[KMSv2] use encDEK, keyID and annotations to generate cache key

It is possible for a KMSv2 plugin to return a static value as Ciphertext
and store the actual encrypted DEK in the annotations. In this case,
using the encDEK will not work. Instead, we are now using a combination
of the encDEK, keyID and annotations to generate the cache key.

Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

aramase commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:
```diff
@@ -861,6 +873,107 @@ func TestEnvelopeLogging(t *testing.T) {
 	}
 }
 
+func TestCacheNotCorrupted(t *testing.T) {
```
This test fails in master as we're only using the `EncryptedDEK` as cache key.

```bash
    /home/anramase/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:926: cipher: message authentication failed
```

aramase commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/cache.go:
```diff
@@ -98,5 +98,11 @@ func (c *simpleCache) keyFunc(s []byte) string {
 // toString performs unholy acts to avoid allocations
 func toString(b []byte) string {
-	return *(*string)(unsafe.Pointer(&b))
+	// unsafe.SliceData relies on cap whereas we want to rely on len
+	if len(b) == 0 {
+		return ""
+	}
+	// Copied from go 1.20.1 strings.Builder.String
+	// https://github.com/golang/go/blob/202a1a57064127c3f19d96df57b9f9586145e21c/src/strings/builder.go#L48
+	return unsafe.String(unsafe.SliceData(b), len(b))
```
These changes have been mirrored from https://github.com/kubernetes/kubernetes/pull/115935

aramase commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:
```diff
@@ -339,3 +355,58 @@ func getRequestInfoFromContext(ctx context.Context) *genericapirequest.RequestIn
+	return b.Bytes()
+}
+
+// toBytes performs unholy acts to avoid allocations
+func toBytes(s string) []byte {
+	// unsafe.StringData is unspecified for the empty string, so we provide a strict interpretation
+	if len(s) == 0 {
+		return nil
+	}
+	// Copied from go 1.20.1 os.File.WriteString
+	// https://github.com/golang/go/blob/202a1a57064127c3f19d96df57b9f9586145e21c/src/os/file.go#L246
+	return unsafe.Slice(unsafe.StringData(s), len(s))
+}
```
These changes have been mirrored from https://github.com/kubernetes/kubernetes/pull/115935

enj commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:
```diff
@@ -339,3 +355,58 @@ func getRequestInfoFromContext(ctx context.Context) *genericapirequest.RequestIn
+
+	// Sort the annotations by key.
+	keys := make([]string, 0, len(annotations))
+	for k := range annotations {
+		keys = append(keys, k)
```
I think the existing code is fine but this makes me feel better:

```suggestion
	for k := range annotations {
		k := k
		keys = append(keys, k)
```

enj commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:
```diff
@@ -137,8 +143,13 @@ func (t *envelopeTransformer) TransformFromStorage(ctx context.Context, data []b
+
 	// Look up the decrypted DEK from cache first
-	transformer := t.cache.get(encryptedObject.EncryptedDEK)
+	transformer := t.cache.get(encryptedObjectCacheKey)
```
The call below to `addTransformerForDecryption` is incorrect - please add a test that catches that :)

aramase replied:
good catch! interesting how the current tests didn't catch it. Let me check!

aramase replied:
Added a new test in `TestEnvelopeCaching` and modified it to do multiple reads to test this. The new table test failed as expected to catch this

```bash
=== RUN   TestEnvelopeCaching
=== RUN   TestEnvelopeCaching/entry_in_cache_should_withstand_plugin_failure
=== RUN   TestEnvelopeCaching/cache_entry_expired_should_not_withstand_plugin_failure
=== RUN   TestEnvelopeCaching/cache_entry_expired_should_work_after_cache_refresh
    /home/anramase/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope_test.go:221: expected 1 decrypt calls, got 10
--- FAIL: TestEnvelopeCaching (0.00s)
    --- PASS: TestEnvelopeCaching/entry_in_cache_should_withstand_plugin_failure (0.00s)
    --- PASS: TestEnvelopeCaching/cache_entry_expired_should_not_withstand_plugin_failure (0.00s)
    --- FAIL: TestEnvelopeCaching/cache_entry_expired_should_work_after_cache_refresh (0.00s)
FAIL
FAIL	k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2	0.005s
```

enj reviewed (changes requested):
Still looking through the tests.

enj commented:
/lgtm
/approve

aramase commented:
/retest
/skip
